week4/exercise.hs

```diff
@@ -0,0 +1,55 @@
+{-# OPTIONS_GHC -Wall #-}
+
+module Week4 where
+
+-- Exercise 1 : Wholemeal programming
+
+fun1' :: [Integer] -> Integer
+fun1' =  product . map (subtract 2) . filter even
+
+fun2' :: Integer -> Integer
+fun2' = sum . filter (even) . takeWhile (/=1) . iterate next
+  where
+    next :: Integer -> Integer
+    next x = if even x then div x 2 else 3 * x + 1
+
+-- Exercise 2 : FoldTree
+
+data Tree a = Leaf 
+    | Node Integer (Tree a) a (Tree a)
+    deriving (Show, Eq)
+
+insert' :: a -> Tree a -> Tree a
+insert' a Leaf = Node 0 Leaf a Leaf
+insert' a (Node _ l b r) =
+    case compare (height l) (height r) of
+        LT -> let x = insert' a l in Node (height x + 1) x b r
+        _  -> let y = insert' a r in Node (height y + 1) l b y
+    where
+        height :: Tree a -> Integer
+        height Leaf = -1
+        height (Node h _ _ _) = h
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insert' Leaf 
+
+-- Exercise 3 : More folds
+
+xor :: [Bool] -> Bool
+xor = foldr1 (/=)
```

Comment:
`foldr1` is a partial function, so this particular implementation will fail on empty lists. Instead, it's better to use`foldr`. You only need to think of a default to make it work with the `/=` operator.